process/transaction/shardProcess.go

```diff
@@ -612,6 +614,35 @@ func (txProc *txProcessor) addFeeAndValueToDest(acntDst state.UserAccountHandler
 	return txProc.accounts.SaveAccount(acntDst)
 }
 
+func (txProc *txProcessor) processRelayedTxV3(
+	tx *transaction.Transaction,
+	relayerAcnt, acntDst state.UserAccountHandler,
+) (vmcommon.ReturnCode, error) {
+	if !txProc.enableEpochsHandler.IsRelayedTransactionsV3FlagEnabled() {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3Disabled)
+	}
+	if tx.GetValue().Cmp(big.NewInt(0)) != 0 {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3ZeroVal)
+	}
+
+	userTx := tx.GetInnerTransaction()
+	if !bytes.Equal(tx.RcvAddr, userTx.SndAddr) {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3BeneficiaryDoesNotMatchReceiver)
+	}
+	if len(userTx.RelayerAddr) == 0 {
```

Comment:
I do not see the need for the user to put a relayerAddress in his own TX. This was used internally, in case of relayedV1 and v2. It was not the user who field this value, but the protocol in case of processing.

Reply:
added on inner txs only, for extra check on relayed v3, as mentioned above

Reply:
add the full check here as well userTx.RelayerAddr = tx.SenderAddr

Reply:
added